pkg/controller/daemon/update.go

```diff
@@ -64,20 +73,311 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
 			break
 		}
 		glog.V(4).Infof("Marking pod %s/%s for deletion", ds.Name, pod.Name)
-		podsToDelete = append(podsToDelete, pod.Name)
+		oldPodsToDelete = append(oldPodsToDelete, pod.Name)
 		numUnavailable++
 	}
-	errors := dsc.syncNodes(ds, podsToDelete, []string{})
-	return utilerrors.NewAggregate(errors)
+	return dsc.syncNodes(ds, oldPodsToDelete, []string{}, hash)
 }
 
-func (dsc *DaemonSetsController) getAllDaemonSetPods(ds *extensions.DaemonSet, nodeToDaemonPods map[string][]*v1.Pod) ([]*v1.Pod, []*v1.Pod) {
+func (dsc *DaemonSetsController) constructHistory(ds *extensions.DaemonSet) (cur *apps.ControllerRevision, old []*apps.ControllerRevision, err error) {
+	var histories []*apps.ControllerRevision
+	var currentHistories []*apps.ControllerRevision
+	histories, err = dsc.controlledHistories(ds)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, history := range histories {
+		// Add the unique label if it's not already added to the history
+		// We use history name instead of computing hash, so that we don't need to worry about hash collision
+		if _, ok := history.Labels[extensions.DefaultDaemonSetUniqueLabelKey]; !ok {
+			var clone interface{}
+			clone, err = api.Scheme.DeepCopy(history)
+			if err != nil {
+				return nil, nil, err
+			}
+			toUpdate := clone.(*apps.ControllerRevision)
+			toUpdate.Labels[extensions.DefaultDaemonSetUniqueLabelKey] = toUpdate.Name
+			history, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Update(toUpdate)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+		// Compare histories with ds to separate cur and old history
+		found := false
+		found, err = Match(&ds.Spec.Template, history)
+		if err != nil {
+			return nil, nil, err
+		}
+		if found {
+			currentHistories = append(currentHistories, history)
+		} else {
+			old = append(old, history)
+		}
+	}
+
+	currRevision := maxRevision(old) + 1
+	switch len(currentHistories) {
+	case 0:
+		// Create a new history if the current one isn't found
+		cur, err = dsc.snapshot(ds, currRevision)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		cur, err = dsc.dedupCurHistories(ds, currentHistories)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Update revision number if necessary
+		if cur.Revision < currRevision {
+			var clone interface{}
+			clone, err = api.Scheme.DeepCopy(cur)
+			if err != nil {
+				return nil, nil, err
+			}
+			toUpdate := clone.(*apps.ControllerRevision)
+			toUpdate.Revision = currRevision
+			_, err = dsc.kubeClient.AppsV1beta1().ControllerRevisions(ds.Namespace).Update(toUpdate)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+	// Label ds with current history's unique label as well
```

Comment:
It's easier for us to check if pods are up-to-date without looking though all history again when calculating DS status